File: backend/src/main/kotlin/com/alpha/service/AuthService.kt
package com.alpha.service

import com.alpha.domain.entity.SessionEntity
import com.alpha.domain.entity.UserEntity
import com.alpha.domain.enums.UserRole
import com.alpha.domain.repository.*
import com.alpha.service.dto.*
import com.alpha.service.exception.*
import org.springframework.security.crypto.password.PasswordEncoder
import org.springframework.stereotype.Service
import org.springframework.transaction.annotation.Transactional
import java.time.OffsetDateTime
import java.util.*
import java.util.regex.Pattern

@Service
@Transactional
class AuthService(
    private val userRepository: UserRepository,
    private val sessionRepository: SessionRepository,
    private val emailVerificationRepository: EmailVerificationRepository,
    private val passwordResetRepository: PasswordResetRepository,
    private val passwordEncoder: PasswordEncoder,
    private val jwtService: JwtService,
    private val emailService: EmailService
) {
    
    private val EMAIL_PATTERN = Pattern.compile("^[A-Za-z0-9+_.-]+@(.+)$")
    
    fun register(request: RegisterRequest, userAgent: String?, ipAddress: String?): AuthResponse {
        // Validate email
        validateEmail(request.email)
        
        // Check if user already exists
        if (userRepository.existsByEmail(request.email)) {
            throw ConflictException("User with this email already exists")
        }
        
        // Create user
        val user = UserEntity().apply {
            email = request.email
            passwordHash = passwordEncoder.encode(request.password)
            name = request.name
            phone = request.phone
            role = request.requiredRole
            emailVerified = false
            isBanned = false
        }
        
        val savedUser = userRepository.save(user)
        
        // Create session
        val session = createSession(savedUser, userAgent, ipAddress)
        
        // Create email verification token
        createEmailVerificationToken(savedUser)
        
        // Send verification email
        emailService.sendVerificationEmail(savedUser.email, savedUser.name ?: "User")
        
        return AuthResponse(
            user = mapToUserResponse(savedUser),
            accessToken = jwtService.generateAccessToken(savedUser),
            refreshToken = session.refreshToken,
            sessionId = session.id.toString()
        )
    }
    
    fun login(request: LoginRequest, userAgent: String?, ipAddress: String?): AuthResponse {
        // Find user by email
        val user = userRepository.findByEmail(request.email)
            ?: throw AuthenticationException("Invalid email or password")
        
        // Check if user is banned
        if (user.isBanned) {
            throw AuthenticationException("Account is banned")
        }
        
        // Verify password
        if (user.passwordHash == null || !passwordEncoder.matches(request.password, user.passwordHash)) {
            throw AuthenticationException("Invalid email or password")
        }
        
        // Create session
        val session = createSession(user, userAgent, ipAddress)
        
        return AuthResponse(
            user = mapToUserResponse(user),
            accessToken = jwtService.generateAccessToken(user),
            refreshToken = session.refreshToken,
            sessionId = session.id.toString()
        )
    }
    
    fun refreshToken(request: RefreshTokenRequest): TokenRefreshResponse {
        // Find session by refresh token
        val session = sessionRepository.findByRefreshToken(request.refreshToken)
            ?: throw AuthenticationException("Invalid refresh token")
        
        // Check if session is expired
        if (session.expiresAt.isBefore(OffsetDateTime.now())) {
            sessionRepository.delete(session)
            throw AuthenticationException("Session expired")
        }
        
        // Find user
        val user = session.user
        
        // Generate new tokens
        val newAccessToken = jwtService.generateAccessToken(user)
        val newRefreshToken = UUID.randomUUID().toString()
        
        // Create new session (instead of updating)
        val newSession = createSession(user, session.userAgent, session.ipAddress)
        
        // Delete old session
        sessionRepository.delete(session)
        
        return TokenRefreshResponse(
            accessToken = newAccessToken,
            refreshToken = newSession.refreshToken
        )
    }
    
    fun logout(sessionId: String, userId: UUID) {
        val session = sessionRepository.findById(UUID.fromString(sessionId))
            .orElseThrow { NotFoundException("Session not found") }
        
        // Verify session belongs to user
        if (session.user.requiredId != userId) {
            throw ForbiddenException("Cannot logout other user's session")
        }
        
        sessionRepository.delete(session)
    }
    
    fun verifyEmail(token: String) {
        val verification = emailVerificationRepository.findByToken(token)
            ?: throw ValidationException("Invalid verification token")
        
        // Check if token is expired
        if (verification.expiresAt.isBefore(OffsetDateTime.now())) {
            emailVerificationRepository.delete(verification)
            throw ValidationException("Verification token expired")
        }
        
        // Update user
        val user = verification.user
        user.emailVerified = true
        userRepository.save(user)
        
        // Delete verification token
        emailVerificationRepository.delete(verification)
    }
    
    fun forgotPassword(email: String) {
        val user = userRepository.findByEmail(email)
            ?: throw NotFoundException("User not found")
        
        // Create password reset token
        val token = UUID.randomUUID().toString()
        val expiresAt = OffsetDateTime.now().plusHours(24)
        
        val passwordReset = passwordResetRepository.save(
            com.alpha.domain.entity.PasswordResetEntity(
                user = user,
                token = token,
                expiresAt = expiresAt
            )
        )
        
        // Send password reset email
        emailService.sendPasswordResetEmail(user.email, user.name ?: "User", token)
    }
    
    fun resetPassword(token: String, newPassword: String) {
        val passwordReset = passwordResetRepository.findByToken(token)
            ?: throw ValidationException("Invalid reset token")
        
        // Check if token is expired
        if (passwordReset.expiresAt.isBefore(OffsetDateTime.now())) {
            passwordResetRepository.delete(passwordReset)
            throw ValidationException("Reset token expired")
        }
        
        // Update user password
        val user = passwordReset.user
        user.passwordHash = passwordEncoder.encode(newPassword)
        userRepository.save(user)
        
        // Delete all user sessions
        sessionRepository.deleteByUserId(user.requiredId)
        
        // Delete password reset token
        passwordResetRepository.delete(passwordReset)
    }
    
    fun getCurrentUser(userId: UUID): UserResponse {
        val user = userRepository.findById(userId)
            .orElseThrow { NotFoundException("User not found") }
        
        return mapToUserResponse(user)
    }
    
    fun updateProfile(userId: UUID, name: String?, phone: String?, avatarUrl: String?) {
        val user = userRepository.findById(userId)
            .orElseThrow { NotFoundException("User not found") }
        
        name?.let { user.name = it }
        phone?.let { user.phone = it }
        avatarUrl?.let { user.avatarUrl = it }
        
        userRepository.save(user)
    }
    
    fun changePassword(userId: UUID, currentPassword: String, newPassword: String) {
        val user = userRepository.findById(userId)
            .orElseThrow { NotFoundException("User not found") }
        
        // Verify current password
        if (user.passwordHash == null || !passwordEncoder.matches(currentPassword, user.passwordHash)) {
            throw ValidationException("Current password is incorrect")
        }
        
        // Update password
        user.passwordHash = passwordEncoder.encode(newPassword)
        userRepository.save(user)
        
        // Delete all user sessions except current one
        // (This forces re-login on all devices for security)
        // Implementation depends on how we track current session
    }
    
    private fun validateEmail(email: String) {
        if (!EMAIL_PATTERN.matcher(email).matches()) {
            throw ValidationException("Invalid email format")
        }
    }
    
    private fun createSession(user: UserEntity, userAgent: String?, ipAddress: String?): SessionEntity {
        val refreshToken = UUID.randomUUID().toString()
        val expiresAt = OffsetDateTime.now().plusDays(7)
        
        return sessionRepository.save(
            SessionEntity(
                user = user,
                token = "", // Not storing access token anymore
                refreshToken = refreshToken,
                userAgent = userAgent,
                ipAddress = ipAddress,
                expiresAt = expiresAt
            )
        )
    }
    
    private fun createEmailVerificationToken(user: UserEntity) {
        val token = UUID.randomUUID().toString()
        val expiresAt = OffsetDateTime.now().plusHours(24)
        
        emailVerificationRepository.save(
            com.alpha.domain.entity.EmailVerificationEntity(
                user = user,
                token = token,
                expiresAt = expiresAt
            )
        )
    }
    
    private fun mapToUserResponse(user: UserEntity): UserResponse {
        return UserResponse(
            id = user.requiredId,
            email = user.email,
            name = user.name,
            phone = user.phone,
            role = user.role,
            emailVerified = user.emailVerified,
            avatarUrl = user.avatarUrl,
            regionId = user.region?.id,
            isBanned = user.isBanned,
            createdAt = user.requiredCreatedAt
        )
    }
}

File: backend-scala-zio/src/main/scala/com/alpha/service/AuthService.scala
package com.alpha.service

import zio.*
import com.alpha.repository.*
import com.alpha.domain.enums.*
import com.alpha.domain.model.*
import com.alpha.dto.*
import com.alpha.provider.*
import org.mindrot.jbcrypt.BCrypt
import java.util.UUID
import java.time.OffsetDateTime

trait AuthService:
  def register(req: RegisterUserRequest): Task[User]
  def login(req: LoginUserRequest, userAgent: Option[String], ipAddress: Option[String]): Task[(User, UserSession)]
  def refreshToken(req: RefreshTokenRequest): Task[(String, String)]
  def logout(sessionId: UUID): Task[Unit]
  def verifyEmail(req: VerifyEmailRequest): Task[Unit]
  def forgotPassword(req: ForgotPasswordRequest): Task[Unit]
  def resetPassword(req: ResetPasswordRequest): Task[Unit]
  def getUser(id: UUID): Task[Option[User]]
  def updateProfile(userId: UUID, req: UpdateProfileRequest): Task[User]
  def changePassword(userId: UUID, req: ChangePasswordRequest): Task[Unit]

object AuthService:
  val layer: ZLayer[
    UserRepository & SessionRepository & EmailVerificationRepository & PasswordResetRepository & TimeProvider & UUIDProvider,
    Nothing,
    AuthService] =
    ZLayer.fromFunction(new AuthServiceImpl(_, _, _, _, _, _))

class AuthServiceImpl(
  userRepo: UserRepository,
  sessionRepo: SessionRepository,
  emailVerificationRepo: EmailVerificationRepository,
  passwordResetRepo: PasswordResetRepository,
  timeProvider: TimeProvider,
  uuidProvider: UUIDProvider
) extends AuthService:

  private def hashPassword(password: String): String                 = BCrypt.hashpw(password, BCrypt.gensalt())
  private def checkPassword(password: String, hash: String): Boolean = BCrypt.checkpw(password, hash)

  override def register(req: RegisterUserRequest): Task[User] =
    for
      exists <- userRepo.existsByEmail(req.email)
      _      <- ZIO.fail(new Exception("Email already exists")).when(exists)
      user    = User(
                  id = uuidProvider.randomUUID(),
                  email = req.email,
                  passwordHash = hashPassword(req.password),
                  name = req.name,
                  phone = req.phone,
                  role = req.role.flatMap(r => UserRole.values.find(_.value == r)).map(_.value).getOrElse("CUSTOMER"),
                  regionId = req.regionId,
                  isActive = true,
                  isBanned = false,
                  emailVerified = false,
                  googleId = None,
                  avatarUrl = None,
                  createdAt = timeProvider.now(),
                  updatedAt = None
                )
      id     <- userRepo.create(user)
    yield user.copy(id = id)

  override def login(
    req: LoginUserRequest,
    userAgent: Option[String],
    ipAddress: Option[String]): Task[(User, UserSession)] =
    for
      userOpt   <- userRepo.findByEmail(req.email)
      user      <- ZIO.fromOption(userOpt).orElseFail(new Exception("Invalid credentials"))
      _         <- ZIO.fail(new Exception("Account is banned")).when(user.isBanned)
      _         <- ZIO.fail(new Exception("Invalid credentials")).when(!checkPassword(req.password, user.passwordHash))
      session    = UserSession(
                     id = uuidProvider.randomUUID(),
                     userId = user.id,
                     refreshToken = uuidProvider.randomUUID().toString,
                     token = uuidProvider.randomUUID().toString,
                     userAgent = userAgent,
                     ipAddress = ipAddress,
                     expiresAt = timeProvider.now().plusDays(7),
                     isRevoked = false,
                     createdAt = timeProvider.now()
                   )
      sessionId <- sessionRepo.create(session)
    yield (user, session.copy(id = sessionId))

  override def refreshToken(req: RefreshTokenRequest): Task[(String, String)] =
    for
      sessionOpt     <- sessionRepo.findByRefreshToken(req.refreshToken)
      session        <- ZIO.fromOption(sessionOpt).orElseFail(new Exception("Invalid refresh token"))
      _              <- ZIO.fail(new Exception("Session revoked")).when(session.isRevoked)
      _              <- ZIO.fail(new Exception("Session expired")).when(session.expiresAt.isBefore(timeProvider.now()))
      userOpt        <- userRepo.findById(session.userId)
      user           <- ZIO.fromOption(userOpt).orElseFail(new Exception("User not found"))
      newAccessToken  = s"access-${user.id}-${uuidProvider.randomUUID()}"
      newRefreshToken = uuidProvider.randomUUID().toString
    yield (newAccessToken, newRefreshToken)

  override def logout(sessionId: UUID): Task[Unit] =
    sessionRepo.revoke(sessionId).unit

  override def verifyEmail(req: VerifyEmailRequest): Task[Unit] =
    for
      verificationOpt <- emailVerificationRepo.findByToken(req.token)
      verification    <- ZIO.fromOption(verificationOpt).orElseFail(new Exception("Invalid verification token"))
      _               <- ZIO.fail(new Exception("Token expired")).when(verification.expiresAt.isBefore(timeProvider.now()))
      _               <- ZIO.fail(new Exception("Token already used")).when(verification.isUsed)
      _               <- emailVerificationRepo.markAsUsed(verification.id)
      userOpt         <- userRepo.findById(verification.userId)
      user            <- ZIO.fromOption(userOpt).orElseFail(new Exception("User not found"))
      _               <- userRepo.update(user.copy(emailVerified = true, updatedAt = Some(timeProvider.now())))
    yield ()

  override def forgotPassword(req: ForgotPasswordRequest): Task[Unit] =
    for
      userOpt <- userRepo.findByEmail(req.email)
      user    <- ZIO.fromOption(userOpt).orElseFail(new Exception("User not found"))
      reset    = PasswordReset(
                   id = uuidProvider.randomUUID(),
                   userId = user.id,
                   token = uuidProvider.randomUUID().toString,
                   expiresAt = timeProvider.now().plusHours(1),
                   isUsed = false,
                   createdAt = timeProvider.now()
                 )
      _       <- passwordResetRepo.create(reset)
    yield ()

  override def resetPassword(req: ResetPasswordRequest): Task[Unit] =
    for
      resetOpt      <- passwordResetRepo.findByToken(req.token)
      reset         <- ZIO.fromOption(resetOpt).orElseFail(new Exception("Invalid reset token"))
      _             <- ZIO.fail(new Exception("Token expired")).when(reset.expiresAt.isBefore(timeProvider.now()))
      _             <- ZIO.fail(new Exception("Token already used")).when(reset.isUsed)
      hashedPassword = hashPassword(req.newPassword)
      userOpt       <- userRepo.findById(reset.userId)
      user          <- ZIO.fromOption(userOpt).orElseFail(new Exception("User not found"))
      _             <- userRepo.update(user.copy(passwordHash = hashedPassword, updatedAt = Some(timeProvider.now())))
      _             <- passwordResetRepo.markAsUsed(reset.id)
    yield ()

  override def getUser(id: UUID): Task[Option[User]] =
    userRepo.findById(id)

  override def updateProfile(userId: UUID, req: UpdateProfileRequest): Task[User] =
    for
      userOpt <- userRepo.findById(userId)
      user    <- ZIO.fromOption(userOpt).orElseFail(new Exception("User not found"))
      updated  = user.copy(
                   name = req.name.orElse(user.name),
                   phone = req.phone.orElse(user.phone),
                   avatarUrl = req.avatarUrl.orElse(user.avatarUrl),
                   updatedAt = Some(timeProvider.now()))
      _       <- userRepo.update(updated)
    yield updated

  override def changePassword(userId: UUID, req: ChangePasswordRequest): Task[Unit] =
    for
      userOpt       <- userRepo.findById(userId)
      user          <- ZIO.fromOption(userOpt).orElseFail(new Exception("User not found"))
      _             <- ZIO.fail(new Exception("Current password is incorrect")).when(!checkPassword(
                         req.currentPassword,
                         user.passwordHash))
      hashedPassword = hashPassword(req.newPassword)
      _             <- userRepo.update(user.copy(passwordHash = hashedPassword, updatedAt = Some(timeProvider.now())))
    yield ()
